crates/ruff_linter/src/noqa.rs

```diff
@@ -275,26 +283,43 @@ pub(crate) fn rule_is_ignored(
     }
 }
 
-/// The file-level exemptions extracted from a given Python file.
+/// A summary of the file-level exemption as extracted from [`BlanketNoqaDirectives`].
 #[derive(Debug)]
-pub(crate) enum FileExemption {
+pub(crate) enum FileExemption<'a> {
     /// The file is exempt from all rules.
     All,
     /// The file is exempt from the given rules.
-    Codes(Vec<NoqaCode>),
+    Codes(Vec<&'a NoqaCode>),
```

Comment:
I left this as a `Vec`, but we could consider using a `HashSet` instead as we're checking if certain `NoqaCode`s are contained in a few places (one being the loop on line 54 in `checkers/noqa.rs`). 

It's probably fine though as I'd expect there will usually only be very few elements here and checking the small `Vec` is likely faster compared to the overhead of hashing every time (and building the `HashSet` in the first place).
We'd also have to derive `Hash` on `NoqaCode`.